test-loop-tests/src/setup/builder.rs

```diff
@@ -467,40 +477,63 @@ impl TestLoopBuilder {
         self
     }
 
+    fn ensure_tempdir(mut self) -> Self {
+        self.test_loop_data_dir.get_or_insert_with(|| tempfile::tempdir().unwrap());
+        self
+    }
+
     fn build_impl(mut self) -> TestLoopEnv {
-        let mut datas = Vec::new();
-        let tempdir =
-            self.test_loop_data_dir.take().unwrap_or_else(|| tempfile::tempdir().unwrap());
-        let network_shared_state = TestLoopNetworkSharedState::new();
-        for idx in 0..self.clients.len() {
-            let account = &self.clients[idx];
-            let is_archival = self.archival_clients.contains(account);
-            let data = self.setup_client(idx, &tempdir, is_archival, &network_shared_state);
-            datas.push(data);
-        }
+        let warmup_pending = self.warmup_pending.clone();
+        self.test_loop.send_adhoc_event("warmup_pending".into(), move |_| {
+            assert!(
+                !warmup_pending.load(Ordering::Relaxed),
+                "Warmup is pending! Call env.warmup() or builder.skip_warmup()"
+            );
+        });
 
-        let env = TestLoopEnv { test_loop: self.test_loop, datas, tempdir };
-        if self.warmup { env.warmup() } else { env }
+        let node_states =
+            (0..self.clients.len()).map(|idx| self.setup_node_state(idx)).collect_vec();
+        let (mut test_loop, shared_state) = self.setup_shared_state();
+        let datas = node_states
+            .into_iter()
+            .map(|node_state| {
+                let account_id = node_state.account_id.clone();
+                setup_client(account_id.as_str(), &mut test_loop, node_state, &shared_state)
+            })
+            .collect_vec();
+
+        TestLoopEnv { test_loop, node_datas: datas, shared_state }
+    }
+
+    fn setup_shared_state(self) -> (TestLoopV2, SharedState) {
+        let shared_state = SharedState {
+            genesis: self.genesis.unwrap(),
+            tempdir: self.test_loop_data_dir.unwrap(),
+            epoch_config_store: self.epoch_config_store.unwrap(),
+            runtime_config_store: self.runtime_config_store,
+            network_shared_state: TestLoopNetworkSharedState::new(),
+            upgrade_schedule: self.upgrade_schedule,
+            chunks_storage: Default::default(),
+            drop_condition_kinds: self.drop_condition_kinds,
```

Comment:
In future PRs I would like to move the drop conditions out of here as it's making things too clunky in the builder.